samples/fpgaperf_counter/fpgaperf_counter.c

```diff
@@ -0,0 +1,639 @@
+			opae_mutex_unlock(res, &fpga_perf_lock);
+			ret = FPGA_EXCEPTION;
+			goto out;
+		}
+		if (opae_mutex_unlock(res, &fpga_perf_lock)) {
+			OPAE_MSG("Failed to unlock perf mutex");
+			ret = FPGA_EXCEPTION;
+			goto out;
+		}
+		if (fpga_Perf_Events(sysfs_perf) != ret)	{
+			OPAE_ERR("Failed to parse fpga perf event");
+			goto out;
+		}
+
+	} else {
+		ret = FPGA_NOT_FOUND;
+		goto out;
+	}
+
+out:
+	globfree(&pglob);
+	return ret;
+}
+
+fpga_result fpgaPerfCounterStartRecord(void)
+{
+	uint64_t loop			= 0;
+	uint64_t inner_loop		= 0;
+	char buf[DFL_PERF_STR_MAX]	= { 0 };
+	struct read_format *rdft	= (struct read_format *) buf;
+	int res				= 0;
+	
+	if (opae_mutex_lock(res, &fpga_perf_lock)) {
+		OPAE_MSG("Failed to lock perf mutex");
+		return FPGA_EXCEPTION;
+	}
+	if (ioctl(g_fpga_perf->perf_events[0].fd, PERF_EVENT_IOC_ENABLE,
+				PERF_IOC_FLAG_GROUP) == -1) {
+		OPAE_ERR("PERF_EVENT_IOC_ENABLE ioctl failed: %s",
+				strerror(errno));
+		goto out;
+	}
+	if (read(g_fpga_perf->perf_events[0].fd, rdft, sizeof(buf)) == -1) {
+		OPAE_ERR("read fpga perf counter failed");
+		goto out;
+	}
+	for (loop = 0; loop < (uint64_t)rdft->nr; loop++) {
+		for (inner_loop = 0; inner_loop < g_fpga_perf->num_perf_events;
+								inner_loop++) {
+			if (rdft->values[loop].id == g_fpga_perf->perf_events[inner_loop].id)
+				g_fpga_perf->perf_events[inner_loop].start_value = rdft->values[loop].value;
+		}
+	}	
+	if (opae_mutex_unlock(res, &fpga_perf_lock)) {
+		OPAE_MSG("Failed to unlock perf mutex");
+		return FPGA_EXCEPTION;
+	}
+	return FPGA_OK;
+out:
+	opae_mutex_unlock(res, &fpga_perf_lock);
+	return FPGA_EXCEPTION;
+}
+
+fpga_result fpgaPerfCounterStopRecord(void)
+{
+	char buf[DFL_PERF_STR_MAX]	= { 0 };
+	uint64_t loop			= 0;
+	uint64_t inner_loop		= 0;
+	struct read_format *rdft 	= (struct read_format *) buf;
+	int res				= 0;
+
+	if (opae_mutex_lock(res, &fpga_perf_lock)) {
+		OPAE_MSG("Failed to lock perf mutex");
+		return FPGA_EXCEPTION;
+	}
+	if (ioctl(g_fpga_perf->perf_events[0].fd, PERF_EVENT_IOC_DISABLE,
+				PERF_IOC_FLAG_GROUP) == -1) {
+		OPAE_ERR("PERF_EVENT_IOC_ENABLE ioctl failed: %s",
+				strerror(errno));
+		goto out;
+	}
+	if (read(g_fpga_perf->perf_events[0].fd, rdft, sizeof(buf)) == -1) {
+		OPAE_ERR("read fpga perf counter failed");
+		goto out;
+	}	
+	for (loop = 0; loop < (uint64_t)rdft->nr; loop++) {
+		for (inner_loop = 0; inner_loop < g_fpga_perf->num_perf_events;
+								inner_loop++) {
+			if (rdft->values[loop].id == g_fpga_perf->perf_events[inner_loop].id)
+				g_fpga_perf->perf_events[inner_loop].stop_value = rdft->values[loop].value;
+		}
+	}
+	if (opae_mutex_unlock(res, &fpga_perf_lock)) {
+		OPAE_MSG("Failed to unlock perf mutex");
+		return FPGA_EXCEPTION;
+	}
+	return FPGA_OK;
+out:
+	opae_mutex_unlock(res, &fpga_perf_lock);
+	return FPGA_EXCEPTION;
+}
+
+fpga_result fpgaPerfCounterPrint(FILE *f)
+{
+	uint64_t loop 	= 0;
+	int res		= 0;
+
+	if (!f)
+		return FPGA_EXCEPTION;
```

Comment:
Should log a message and return FPGA_INVALID_PARAM

Reply:
changed as per the comment